cvmfs/sqlitemem.cc

```diff
@@ -0,0 +1,508 @@
+    }
+  }
+  return NULL;
+}
+
+
+bool MemoryManager::LookasideBufferArena::IsEmpty() {
+  for (unsigned i = 0; i < kNoBitmaps; ++i) {
+    if (~freemap_[i] != 0)
+      return false;
+  }
+  return true;
+}
+
+
+bool MemoryManager::LookasideBufferArena::Contains(void *buffer) {
+  if ((buffer == NULL) || (buffer < arena_))
+    return false;
+  return (reinterpret_cast<char *>(buffer) - reinterpret_cast<char *>(arena_)) <
+         kArenaSize;
+}
+
+
+MemoryManager::LookasideBufferArena::LookasideBufferArena()
+  : arena_(sxmmap(kArenaSize))
+{
+  // All buffers unused, i.e. all bits set
+  memset(freemap_, 0xFF, kNoBitmaps * sizeof(int));
+}
+
+
+MemoryManager::LookasideBufferArena::~LookasideBufferArena() {
+  sxunmap(arena_, kArenaSize);
+}
+
+
+void MemoryManager::LookasideBufferArena::PutBuffer(void *buffer) {
+  ptrdiff_t nbuffer =
+    (reinterpret_cast<char *>(buffer) - reinterpret_cast<char *>(arena_))
+    / kBufferSize;
+  assert(nbuffer < kBuffersPerArena);
+  const int nfreemap = nbuffer / (sizeof(int) * 8);
+  freemap_[nfreemap] |= 1 << (nbuffer % (sizeof(int) * 8));
+}
+
+
+//------------------------------------------------------------------------------
+
+
+MemoryManager *MemoryManager::instance_ = NULL;
+
+
+/**
+ * Sqlite ensures that size > 0.
+ */
+void *MemoryManager::xMalloc(int size) {
+  return instance_->GetMemory(size);
+}
+
+
+/**
+ * Sqlite ensures that ptr != NULL.
+ */
+void MemoryManager::xFree(void *ptr) {
+  instance_->PutMemory(ptr);
+}
+
+
+/**
+ * Sqlite ensures that ptr != NULL and new_size > 0.
+ */
+void *MemoryManager::xRealloc(void *ptr, int new_size) {
+  void *new_ptr = xMalloc(new_size);
+  memcpy(new_ptr, ptr, new_size);
+  xFree(ptr);
+  return new_ptr;
+}
+
+
+/**
+ * Sqlite ensures that ptr != NULL.
+ */
+int MemoryManager::xSize(void *ptr) {
+  return instance_->GetMemorySize(ptr);
+}
+
+
+int MemoryManager::xRoundup(int size) {
+  return size;
+}
+
+
+int MemoryManager::xInit(void *app_data __attribute__((unused))) {
+  return SQLITE_OK;
+}
+
+
+void MemoryManager::xShutdown(void *app_data __attribute__((unused))) {
+}
+
+
+void MemoryManager::AssignGlobalArenas() {
+  if (assigned_) return;
+  int retval;
+
+  retval = sqlite3_config(SQLITE_CONFIG_SCRATCH, scratch_memory_,
+                          kScratchSlotSize, kScratchNoSlots);
+  assert(retval == SQLITE_OK);
+
+  retval = sqlite3_config(SQLITE_CONFIG_PAGECACHE, page_cache_memory_,
+                          kPageCacheSlotSize, kPageCacheNoSlots);
+  assert(retval == SQLITE_OK);
+
+  retval = sqlite3_config(SQLITE_CONFIG_GETMALLOC, &sqlite3_mem_vanilla_);
+  assert(retval == SQLITE_OK);
+  retval = sqlite3_config(SQLITE_CONFIG_MALLOC, &mem_methods_);
+  assert(retval == SQLITE_OK);
+
+  assigned_ = true;
+}
+
+
+/**
+ * Needs to be the first operation on an opened sqlite database.  Returns the
+ * location of the buffer.
+ */
+void *MemoryManager::AssignLookasideBuffer(void *db) {
```

Comment:
why `void *db` when it `reinterpret_cast<>`s it into an `sqlite3 *` right after entering the method?